docs: securing mm2 (#6535)
strimzi/strimzi-kafka-operator: documentation/modules/configuring/proc-config-mirrormaker2-securing-connection.adoc

* xref:con-securing-kafka-authorization-str[Supported authorization options for a Kafka cluster]

You configure listeners for secure connections to Kafka clusters.
You can use internal listeners for connections within a Kubernetes cluster and external listeners for connections outside the Kubernetes cluster.

The security options implemented for the source and target Kafka clusters must be compatible with the security options implemented for MirrorMaker 2.0.

NOTE: In this procedure, the certificates generated by the Cluster Operator are used, but you can replace them by xref:installing-your-own-ca-certificates-str[installing your own certificates].
You can also configure your listener to xref:kafka-listener-certificates-str[use a Kafka listener certificate managed by an external Certificate Authority].

.Before you start
Before starting this procedure, take a look at the link:{BookURLDeploying}#deploy-examples-{context}[example configuration files^] provided by Strimzi.
They include examples for securing a deployment of MirrorMaker 2.0 using TLS or SCRAM-SHA-512 authentication.
The examples specify internal listeners for connecting within a Kubernetes cluster.

> The examples also show the ACLs needed by MirrorMaker 2.0 to allow operations on the source and target Kafka clusters.

Author: Not sure if we want to list the ACLs required by the connectors in the doc

.Prerequisites

* Strimzi is running
* Source and target Kafka clusters are available

.Procedure

. Configure two `Kafka` resources, one to secure the source Kafka cluster and one to secure the target Kafka cluster.
+
You can add listener configuration for authentication and enable authorization.
+
In this example, an internal listener is configured for a Kafka cluster with TLS encryption and authentication.
Kafka `simple` authorization is enabled.
+

> .Example source Kafka cluster configuration with TLS authentication

Author: I've added both.

kind: Kafka
metadata:
  name: my-source-cluster
spec:
  kafka:
    version: {DefaultKafkaVersion}
    replicas: 1
    listeners:
      - name: tls
        port: 9093
        type: internal
        tls: true
        authentication:
          type: tls
    authorization:
      type: simple
    config:
      offsets.topic.replication.factor: 1
      transaction.state.log.replication.factor: 1
      transaction.state.log.min.isr: 1
      default.replication.factor: 1
      min.insync.replicas: 1
      inter.broker.protocol.version: "{DefaultInterBrokerVersion}"
    storage:
      type: jbod
      volumes:
      - id: 0
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
  zookeeper:
    replicas: 1
    storage:
      type: persistent-claim
      size: 100Gi
      deleteClaim: false
  entityOperator:
    topicOperator: {}
    userOperator: {}
----

. Create or update the `Kafka` resources.
+
[source,shell,subs=+quotes]
----
kubectl apply -f _<kafka_configuration_file>_
----
+
The Cluster Operator creates the listeners and sets up the cluster and client certificate authority (CA) certificates to enable authentication within the Kafka cluster.
+

> The certificates are created with the same name as the Kafka resource.

Reviewer: I'm not sure this makes sense. I guess you  might want to say something like

```suggestion
The certificates are created in the secret `<cluster-name>-cluster-ca-cert`.
```

The certificates do not really have a name and the Secret name differs.

. Configure two `KafkaUser` resources, one for a user of the source Kafka cluster and one for a user of the target Kafka cluster.
+
The users provide the credentials required for MirrorMaker to connect to the source and target Kafka clusters.
+
--
.. Configure the same authentication and authorization types as the corresponding source and target Kafka cluster.
For example, if you used `tls` authentication and `simple` authorization type in the `Kafka` configuration for the source Kafka cluster,
use the same in the `KafkaUser` configuration.

.. Configure the ACLs needed by MirrorMaker 2.0 to allow operations on the source and target Kafka clusters.
+
The ACLs are used by the internal MirrorMaker connectors, and by the underlying Kafka Connect framework.
--
+
.Example user configuration for TLS client authentication
[source,yaml,subs="attributes+"]
----

> apiVersion: {KafkaUserApiVersion}

Reviewer: Same as above, the users are actually different. So listing only one is weird and can be confusing. I would either list none or both.

Author: I've added both

      transaction.state.log.replication.factor: 1
      transaction.state.log.min.isr: 1
      default.replication.factor: 1
      min.insync.replicas: 1
      inter.broker.protocol.version: "{DefaultInterBrokerVersion}"
    storage:
      type: jbod
      volumes:
      - id: 0
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
  zookeeper:
    replicas: 1
    storage:
      type: persistent-claim
      size: 100Gi
      deleteClaim: false
  entityOperator:
    topicOperator: {}
    userOperator: {}
----
+
.Example target Kafka cluster configuration with TLS authentication
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: Kafka
metadata:
  name: my-target-cluster
spec:
  kafka:
    version: {DefaultKafkaVersion}
    replicas: 1
    listeners:
      - name: tls
        port: 9093
        type: internal
        tls: true
        authentication:
          type: tls
    authorization:
      type: simple
    config:
      offsets.topic.replication.factor: 1
      transaction.state.log.replication.factor: 1
      transaction.state.log.min.isr: 1
      default.replication.factor: 1
      min.insync.replicas: 1

>       inter.broker.protocol.version: "3.1"

Reviewer: ```suggestion
      inter.broker.protocol.version: "{DefaultInterBrokerVersion}"
```

    topicOperator: {}
    userOperator: {}
----
+
.Example target Kafka cluster configuration with TLS authentication
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: Kafka
metadata:
  name: my-target-cluster
spec:
  kafka:
    version: {DefaultKafkaVersion}
    replicas: 1
    listeners:
      - name: tls
        port: 9093
        type: internal
        tls: true
        authentication:
          type: tls
    authorization:
      type: simple
    config:
      offsets.topic.replication.factor: 1
      transaction.state.log.replication.factor: 1
      transaction.state.log.min.isr: 1
      default.replication.factor: 1
      min.insync.replicas: 1
      inter.broker.protocol.version: "3.1"
    storage:
      type: jbod
      volumes:
        - id: 0
          type: persistent-claim
          size: 100Gi
          deleteClaim: false
  zookeeper:
    replicas: 1
    storage:
      type: persistent-claim
      size: 100Gi
      deleteClaim: false
  entityOperator:

>     userOperator: {}

Reviewer: Since you have different namespaces, you can have `topicOperator: {}` here.

      # MirrorCheckpointConnector
      - resource:
          type: cluster
        operation: Describe
      - resource:
          type: topic
          name: my-source-cluster.checkpoints.internal
        operation: Create
      - resource:
          type: topic
          name: my-source-cluster.checkpoints.internal
        operation: Describe
      - resource:
          type: topic
          name: my-source-cluster.checkpoints.internal
        operation: Write
      - resource: # Needed for every group for which the offset is synced
          type: group
          name: "*"
        operation: Read
      - resource: # Needed for every group for which the offset is synced
          type: group
          name: "*"
        operation: Describe
      - resource: # Needed for every topic which is mirrored
          type: topic
          name: "*"
        operation: Read
      # MirrorHeartbeatConnector
      - resource:
          type: topic
          name: heartbeats
        operation: Create
      - resource:
          type: topic
          name: heartbeats
        operation: Describe
      - resource:
          type: topic
          name: heartbeats
        operation: Write
----

> +

Reviewer: We need to make sure it is clear to the users that each user has to be create in different namespace.

      topicsPattern: ".*"
      groupsPattern: ".*"
----
<1> The TLS certificates for the source Kafka cluster. If they are in a separate namespace, copy the cluster secrets from the namespace of the Kafka cluster.
<2> The user authentication for accessing the source Kafka cluster using the xref:type-KafkaClientAuthenticationTls-reference[TLS mechanism].
<1> The TLS certificates for the target Kafka cluster.
<2> The user authentication for accessing the target Kafka cluster.

> . Create or update the `KafkaMirrorMaker2` resource.

Reviewer: We should probably point the user into which namespace should this be deployed? (the namespace of the target cluster)
